File: 116-populating-next-right-pointers-in-each-node/116-populating-next-right-pointers-in-each-node.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        Queue<Node> q = new LinkedList<>();
        q.add(root);
        while (q.size() > 0){
            Node prec = null;
            Node last = null;
            int len = q.size();
            for(int i = 0; i < len; i++){
                last = q.poll();
                if(last != null){
                    q.add(last.left);
                    q.add(last.right);
                }
                if(prec != null){
                    prec.next = last;
                }
                prec = last;
            }
            if(last != null){
                last.next = null;
            }
        }
        return root;
    }
}

File: 116-populating-next-right-pointers-in-each-node/116-populating-next-right-pointers-in-each-node.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        q = []
        q.append(root)
        while q:
            prec = None
            last = None
            size = len(q)
            for i in range(size):
                last = q.pop(0)
                if last != None:
                    q.append(last.left)
                    q.append(last.right)
                
                if prec != None:
                    prec.next = last
                
                prec = last
            
            if last != None:
                last.next = None
            
        return root